crates/subspace-archiving/src/archiver/incremental_record_commitments.rs

```diff
@@ -0,0 +1,194 @@
+    incremental_record_commitments: &mut IncrementalRecordCommitmentsState,
+    segment: &Segment,
+    full: bool,
+    record_size: usize,
+) {
+    segment.encode_to(&mut IncrementalRecordCommitmentsProcessor::new(
+        incremental_record_commitments,
+        record_size,
+        full,
+    ));
+}
+
+/// Processor is hidden to not expose unnecessary implementation details (like `Output` trait
+/// implementation)
+struct IncrementalRecordCommitmentsProcessor<'a> {
+    /// Processed bytes in the segment so far
+    processed: usize,
+    /// Record commitments already created
+    incremental_record_commitments: &'a mut IncrementalRecordCommitmentsState,
+    /// Record size
+    record_size: usize,
+    /// Whether segment is full or partial
+    full: bool,
+    /// Intermediate hashing state that computes Blake2-256-254.
+    ///
+    /// See [`subspace_core_primitives::crypto::blake2b_256_254_hash`] for details.
+    hashing_state: Blake2b<U32>,
+}
+
+impl<'a> Drop for IncrementalRecordCommitmentsProcessor<'a> {
+    fn drop(&mut self) {
+        if self.full {
+            // How many bytes to read before the start of the next record
+            let bytes_until_full_record = self.bytes_until_full_record();
+            if bytes_until_full_record > 0 {
+                // This is fine since we'll have at most a few iterations and allocation is less
+                // desirable than a loop here
+                for _ in 0..bytes_until_full_record {
+                    self.update_hashing_state(&[0]);
+                }
+                self.finish_hash();
+            }
+        }
+    }
+}
+
+impl<'a> Output for IncrementalRecordCommitmentsProcessor<'a> {
+    fn write(&mut self, bytes: &[u8]) {
+        // Try to finish last partial record if possible
+        let bytes = {
+            let bytes_until_full_record =
+                ((self.processed / self.record_size) + 1) * self.record_size - self.processed;
+            let (remaining_record_bytes, bytes) =
+                bytes.split_at(if bytes.len() >= bytes_until_full_record {
+                    bytes_until_full_record
+                } else {
+                    bytes.len()
+                });
+
+            self.update_hashing_state(remaining_record_bytes);
+
+            if remaining_record_bytes.len() == bytes_until_full_record {
+                self.finish_hash();
+            }
+
+            bytes
+        };
+
+        // Continue processing records (full and partial) from remaining data
+        bytes.chunks(self.record_size).for_each(|record| {
+            self.update_hashing_state(record);
+
+            // Store hashes of full records
+            if record.len() == self.record_size {
+                self.finish_hash();
+            }
+        });
+    }
+}
+
+impl<'a> IncrementalRecordCommitmentsProcessor<'a> {
+    fn new(
+        incremental_record_commitments: &'a mut IncrementalRecordCommitmentsState,
+        record_size: usize,
```

Comment:
`record_size` IMHO should be protected by `NonZeroUsize` struct because it used in division extensibly.

Reply:
I agree it would be great, but I think we should move this to a different PR. We need to convert to NonZeroU32/NonZeroUsize in many more places across the whole codebase. Also this is an internal implementation and Acrhiver's constructor will check for sane `record_root` value, so we should be fine here.